storage/remote/queue_manager.go

```diff
@@ -1028,7 +1032,90 @@ func (s *shards) enqueue(ref chunks.HeadSeriesRef, data interface{}) bool {
 	}
 }
 
-func (s *shards) runShard(ctx context.Context, shardID int, queue chan interface{}) {
+type queue struct {
+	batch      []interface{}
+	batchQueue chan []interface{}
+
+	// Since we know there is a limited number of batches out, using a stack is
+	// easy and persistent across GC compared to a sync.Pool
+	batchPool [][]interface{}
+	poolMux   sync.Mutex
```

Comment:
Could we have a comment saying what this Mutex covers?

Reply:
Yep, added.